3-enrich/scala-common-enrich/src/main/scala/com.snowplowanalytics.snowplow.enrich/common/enrichments/registry/PiiPseudonymizerEnrichment.scala

```diff
@@ -0,0 +1,222 @@
+      }.sequenceU
+    } yield PiiPseudonymizerEnrichment(piiFieldList)
+  }.leftMap(getProcessingMessage(_)).toValidationNel
+
+  private def getProcessingMessage(s: String): ProcessingMessage = {
+    val pm = new ProcessingMessage()
+    pm.setMessage(s)
+    pm
+  }
+
+  private def matchesSchema(config: JValue, schemaKey: SchemaKey): Validation[String, JValue] =
+    if (supportedSchema matches schemaKey) {
+      config.success
+    } else {
+      ("Schema key %s is not supported. A '%s' enrichment must have schema '%s'.")
+        .format(schemaKey, supportedSchema.name, supportedSchema)
+        .fail
+    }
+}
+
+/**
+ * The PiiPseudonymizerEnrichment runs after all other enrichments to find fields that are configured as PII and apply
+ * some anonymization (currently only psudonymizantion) on them. Currently a single strategy for all the fields is
+ * supported due to the config format, and there is only one implemented strategy, however the enrichment supports a
+ * strategy per field configuration.
+ *
+ * The user may specify two types of fields POJO or JSON. A POJO field is effectively a scalar field in the
+ * EnrichedEvent, whereas a JSON is a "context" formatted field (a JSON string in "contexts" field in enriched event)
+ *
+ * @param fieldList a lits of configured PiiFields
+ */
+case class PiiPseudonymizerEnrichment(fieldList: List[PiiField]) extends Enrichment {
+  override val version: DefaultArtifactVersion              = new DefaultArtifactVersion("99999.66666.33333")
+  def transformer(transformMap: TransformMap): TransformMap = transformMap ++ fieldList.map(_.transformer(transformMap)).reduce(_ ++ _)
+}
+
+/**
+ * Specifies a field in POJO and the trategy that should be applied to it.
+ * @param strategy
+ * @param fieldName
+ */
+final case class PiiPojo(strategy: PiiStrategy, fieldName: String) extends PiiField {
+  override def applyStrategy(fieldValue: String): String = strategy.scramble(fieldValue)
+}
+
+/**
+ * Specifies a strategy to use, a field (should be "contexts") where the JSON can be found, a schema criterion to
+ * discriminate which contexts to apply this strategy to, and a json path within the contexts where this strategy will
+ * be apllied (the path may correspond to multiple fields).
+ *
+ * @param strategy
+ * @param fieldName
+ * @param schemaCriterion
+ * @param jsonPath
+ */
+final case class PiiJson(strategy: PiiStrategy, fieldName: String, schemaCriterion: SchemaCriterion, jsonPath: String) extends PiiField {
+  override def applyStrategy(fieldValue: String): String = {
+    val jv = parse(fieldValue)
+    val jv1 = jv.transformField {
+      case JField("data", contents) =>
+        ("data", contents.transform {
+          case ja: JArray => {
+            ja.transform {
+              case JObject(List(("schema", JString(schema)), ("data", jobject)))
+                  if (SchemaKey.parse(schema).flatMap(schemaCriterion.matches(_).success).getOrElse(false)) => {
```

Comment:
Why `flatMap` AND `x.success`. This is just `map`.

Reply:
good point